File: scripts/law_docx_parser.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import hashlib
import re

from docx import Document


ARTICLE_RE = re.compile(r"^(第[一二三四五六七八九十百千万零〇两]+条)\s*(.*)$")
BOOK_RE = re.compile(r"^第[一二三四五六七八九十百千万零〇两]+编\s+")
CHAPTER_RE = re.compile(r"^第[一二三四五六七八九十百千万零〇两]+章\s+")
SECTION_RE = re.compile(r"^第[一二三四五六七八九十百千万零〇两]+节\s+")
PASS_DATE_RE = re.compile(r"(\d{4})年(\d{1,2})月(\d{1,2})日")


@dataclass(frozen=True)
class LawArticle:
    article_no: str
    article_order: int
    content: str
    book_title: str | None
    chapter_title: str | None
    section_title: str | None
    full_path: str
    content_hash: str


@dataclass(frozen=True)
class ParsedLawDocument:
    title: str
    law_type: str
    publish_date: str | None
    source_file_name: str
    articles: list[LawArticle]

    def article_by_no(self, article_no: str) -> LawArticle:
        for article in self.articles:
            if article.article_no == article_no:
                return article
        raise KeyError(article_no)


@dataclass
class _ArticleDraft:
    article_no: str
    article_order: int
    parts: list[str]
    book_title: str | None
    chapter_title: str | None
    section_title: str | None

    def append(self, text: str) -> None:
        if text:
            self.parts.append(text)

    def build(self, document_title: str) -> LawArticle:
        content = "\n".join(self.parts).strip()
        path_parts = [
            document_title,
            self.book_title,
            self.chapter_title,
            self.section_title,
            self.article_no,
        ]
        full_path = " > ".join(part for part in path_parts if part)
        return LawArticle(
            article_no=self.article_no,
            article_order=self.article_order,
            content=content,
            book_title=self.book_title,
            chapter_title=self.chapter_title,
            section_title=self.section_title,
            full_path=full_path,
            content_hash=hashlib.sha256(content.encode("utf-8")).hexdigest(),
        )


def parse_law_docx(path: str | Path, law_type: str = "法律") -> ParsedLawDocument:
    source_path = Path(path)
    paragraphs = _read_paragraphs(source_path)
    if len(paragraphs) < 3:
        raise ValueError(f"document has too few paragraphs: {source_path}")

    title = paragraphs[0]
    publish_date = _extract_publish_date(paragraphs[1])
    body_start = _find_body_start(paragraphs)

    book_title: str | None = None
    chapter_title: str | None = None
    section_title: str | None = None
    drafts: list[_ArticleDraft] = []
    current: _ArticleDraft | None = None

    for text in paragraphs[body_start:]:
        if BOOK_RE.match(text):
            book_title = _normalize_heading(text)
            chapter_title = None
            section_title = None
            continue
        if CHAPTER_RE.match(text):
            chapter_title = _normalize_heading(text)
            section_title = None
            continue
        if SECTION_RE.match(text):
            section_title = _normalize_heading(text)
            continue

        article_match = ARTICLE_RE.match(text)
        if article_match:
            article_no, article_content = article_match.groups()
            current = _ArticleDraft(
                article_no=article_no,
                article_order=len(drafts) + 1,
                parts=[article_content.strip()],
                book_title=book_title,
                chapter_title=chapter_title,
                section_title=section_title,
            )
            drafts.append(current)
            continue

        if current is not None:
            current.append(text)

    articles = [draft.build(title) for draft in drafts]
    return ParsedLawDocument(
        title=title,
        law_type=law_type,
        publish_date=publish_date,
        source_file_name=source_path.name,
        articles=articles,
    )


def _read_paragraphs(path: Path) -> list[str]:
    document = Document(path)
    return [_normalize_text(p.text) for p in document.paragraphs if _normalize_text(p.text)]


def _normalize_text(text: str) -> str:
    return re.sub(r"\s+", " ", text.replace("\u3000", " ")).strip()


def _normalize_heading(text: str) -> str:
    parts = text.split(" ", 1)
    if len(parts) == 1:
        return text
    prefix, title = parts
    return f"{prefix} {title.replace(' ', '')}"


def _extract_publish_date(text: str) -> str | None:
    match = PASS_DATE_RE.search(text)
    if not match:
        return None
    year, month, day = match.groups()
    return f"{int(year):04d}-{int(month):02d}-{int(day):02d}"


def _find_body_start(paragraphs: list[str]) -> int:
    first_article_index = next(
        (index for index, text in enumerate(paragraphs) if ARTICLE_RE.match(text)),
        None,
    )
    if first_article_index is None:
        raise ValueError("no article paragraphs found")

    start = first_article_index
    while start > 0 and _is_hierarchy_heading(paragraphs[start - 1]):
        start -= 1
    return start


def _is_hierarchy_heading(text: str) -> bool:
    return bool(BOOK_RE.match(text) or CHAPTER_RE.match(text) or SECTION_RE.match(text))

File: backend/src/main/java/com/law4x/law/parser/LawDocxParser.java
package com.law4x.law.parser;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.poi.xwpf.usermodel.XWPFDocument;

public class LawDocxParser {

    private static final Pattern ARTICLE_PATTERN = Pattern.compile("^(第[一二三四五六七八九十百千万零〇两]+条)\\s*(.*)$");
    private static final Pattern BOOK_PATTERN = Pattern.compile("^第[一二三四五六七八九十百千万零〇两]+编\\s+.*$");
    private static final Pattern CHAPTER_PATTERN = Pattern.compile("^第[一二三四五六七八九十百千万零〇两]+章\\s+.*$");
    private static final Pattern SECTION_PATTERN = Pattern.compile("^第[一二三四五六七八九十百千万零〇两]+节\\s+.*$");
    private static final Pattern PASS_DATE_PATTERN = Pattern.compile("(\\d{4})年(\\d{1,2})月(\\d{1,2})日");

    public ParsedLawDocument parse(Path path) throws IOException {
        return parse(path, "法律");
    }

    public ParsedLawDocument parse(Path path, String lawType) throws IOException {
        List<String> paragraphs = readParagraphs(path);
        if (paragraphs.size() < 3) {
            throw new IllegalArgumentException("document has too few paragraphs: " + path);
        }

        String title = paragraphs.get(0);
        String publishDate = extractPublishDate(paragraphs.get(1));
        int bodyStart = findBodyStart(paragraphs);

        String bookTitle = null;
        String chapterTitle = null;
        String sectionTitle = null;
        List<ArticleDraft> drafts = new ArrayList<>();
        ArticleDraft current = null;

        for (int index = bodyStart; index < paragraphs.size(); index++) {
            String text = paragraphs.get(index);

            if (BOOK_PATTERN.matcher(text).matches()) {
                bookTitle = normalizeHeading(text);
                chapterTitle = null;
                sectionTitle = null;
                continue;
            }
            if (CHAPTER_PATTERN.matcher(text).matches()) {
                chapterTitle = normalizeHeading(text);
                sectionTitle = null;
                continue;
            }
            if (SECTION_PATTERN.matcher(text).matches()) {
                sectionTitle = normalizeHeading(text);
                continue;
            }

            Matcher articleMatcher = ARTICLE_PATTERN.matcher(text);
            if (articleMatcher.matches()) {
                current = new ArticleDraft(
                        articleMatcher.group(1),
                        drafts.size() + 1,
                        bookTitle,
                        chapterTitle,
                        sectionTitle
                );
                current.append(articleMatcher.group(2).trim());
                drafts.add(current);
                continue;
            }

            if (current != null) {
                current.append(text);
            }
        }

        List<LawArticle> articles = drafts.stream()
                .map(draft -> draft.build(title))
                .toList();

        return new ParsedLawDocument(
                title,
                lawType,
                publishDate,
                path.getFileName().toString(),
                articles
        );
    }

    private static List<String> readParagraphs(Path path) throws IOException {
        try (InputStream inputStream = Files.newInputStream(path);
             XWPFDocument document = new XWPFDocument(inputStream)) {
            return document.getParagraphs().stream()
                    .map(paragraph -> normalizeText(paragraph.getText()))
                    .filter(text -> !text.isBlank())
                    .toList();
        }
    }

    private static String normalizeText(String text) {
        return text.replace('\u3000', ' ')
                .replaceAll("\\s+", " ")
                .trim();
    }

    private static String normalizeHeading(String text) {
        String[] parts = text.split(" ", 2);
        if (parts.length == 1) {
            return text;
        }
        return parts[0] + " " + parts[1].replace(" ", "");
    }

    private static String extractPublishDate(String text) {
        Matcher matcher = PASS_DATE_PATTERN.matcher(text);
        if (!matcher.find()) {
            return null;
        }
        return "%04d-%02d-%02d".formatted(
                Integer.parseInt(matcher.group(1)),
                Integer.parseInt(matcher.group(2)),
                Integer.parseInt(matcher.group(3))
        );
    }

    private static int findBodyStart(List<String> paragraphs) {
        for (int index = 0; index < paragraphs.size(); index++) {
            if (ARTICLE_PATTERN.matcher(paragraphs.get(index)).matches()) {
                int start = index;
                while (start > 0 && isHierarchyHeading(paragraphs.get(start - 1))) {
                    start--;
                }
                return start;
            }
        }
        throw new IllegalArgumentException("no article paragraphs found");
    }

    private static boolean isHierarchyHeading(String text) {
        return BOOK_PATTERN.matcher(text).matches()
                || CHAPTER_PATTERN.matcher(text).matches()
                || SECTION_PATTERN.matcher(text).matches();
    }

    private record ArticleDraft(
            String articleNo,
            int articleOrder,
            String bookTitle,
            String chapterTitle,
            String sectionTitle,
            List<String> parts
    ) {
        ArticleDraft(String articleNo, int articleOrder, String bookTitle, String chapterTitle, String sectionTitle) {
            this(articleNo, articleOrder, bookTitle, chapterTitle, sectionTitle, new ArrayList<>());
        }

        void append(String text) {
            if (text != null && !text.isBlank()) {
                parts.add(text);
            }
        }

        LawArticle build(String documentTitle) {
            String content = String.join("\n", parts).trim();
            List<String> pathParts = new ArrayList<>();
            pathParts.add(documentTitle);
            addIfPresent(pathParts, bookTitle);
            addIfPresent(pathParts, chapterTitle);
            addIfPresent(pathParts, sectionTitle);
            pathParts.add(articleNo);

            return new LawArticle(
                    articleNo,
                    articleOrder,
                    content,
                    bookTitle,
                    chapterTitle,
                    sectionTitle,
                    String.join(" > ", pathParts),
                    sha256(content)
            );
        }

        private static void addIfPresent(List<String> values, String value) {
            if (value != null && !value.isBlank()) {
                values.add(value);
            }
        }
    }

    private static String sha256(String value) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return HexFormat.of().formatHex(digest.digest(value.getBytes(StandardCharsets.UTF_8)));
        } catch (NoSuchAlgorithmException exception) {
            throw new IllegalStateException("SHA-256 is not available", exception);
        }
    }
}
